src/extra/gfx/view.cpp

```diff
@@ -0,0 +1,167 @@
+  static SHTypesInfo outputTypes() { return CoreInfo::NoneType; }
+
+  static SHParametersInfo parameters() {
+    static Parameters params{};
+    return params;
+  }
+
+  void cleanup() { baseConsumerCleanup(); }
+
+  void warmup(SHContext *context) { baseConsumerWarmup(context); }
+
+  SHVar activate(SHContext *shContext, const SHVar &input) {
+    ViewStack::Item viewItem{};
+    input::InputStack::Item inputItem;
+
+    // (optional) set viewport rectangle
+    auto &inputTable = input.payload.tableValue;
+    SHVar viewportVar{};
+    if (getFromTable(shContext, inputTable, "Viewport", viewportVar)) {
+      auto &v = viewportVar.payload.float4Value;
+      // Read Float4 rect as (X0, Y0, X1, Y1)
+      viewItem.viewport = Rect::fromCorners(v[0], v[1], v[2], v[3]);
+    }
+
+    // (optional) set render target
+    SHVar rtVar{};
+    if (getFromTable(shContext, inputTable, "Target", rtVar)) {
+      SHRenderTarget *renderTarget = varAsObjectChecked<SHRenderTarget>(rtVar, Types::RenderTarget);
+      viewItem.renderTarget = renderTarget->renderTarget;
+    }
+
+    // (optional) Push window region for input
+    SHVar windowRegionVar{};
+    if (getFromTable(shContext, inputTable, "WindowRegion", windowRegionVar)) {
+      // Read Float4 rect as (X0, Y0, X1, Y1)
+      auto &v = windowRegionVar.payload.float4Value;
+      int4 region(v[0], v[1], v[2], v[3]);
+
+      // This is used for translating cursor inputs from window to view space
+      // TODO: Adjust relative to parent region, for now always assume this is in global window coordinates
+      inputItem.windowMapping = input::WindowSubRegion{
+          .region = region,
+      };
+
+      // Safe to automatically set reference size based on region
+      float2 regionSize = float2(region.z - region.x, region.w - region.y);
+      viewItem.referenceSize = int2(linalg::floor(float2(regionSize)));
+    }
+
+    auto &mwg = getMainWindowGlobals();
+    ViewStack &viewStack = mwg.renderer->getViewStack();
+    input::InputStack &inputStack = mwg.inputStack;
+
+    viewStack.push(std::move(viewItem));
+    inputStack.push(std::move(inputItem));
+
+    return Var::Empty;
+  }
+};
+
+struct PopViewShard : public BaseConsumer {
+  static SHTypesInfo inputTypes() { return CoreInfo::NoneType; }
+  static SHTypesInfo outputTypes() { return CoreInfo::NoneType; }
+
+  static SHParametersInfo parameters() {
+    static Parameters params{};
+    return params;
+  }
+
+  void cleanup() { baseConsumerCleanup(); }
+  void warmup(SHContext *context) { baseConsumerWarmup(context); }
+  SHVar activate(SHContext *shContext, const SHVar &input) {
+    auto &mwg = getMainWindowGlobals();
+    ViewStack &viewStack = mwg.renderer->getViewStack();
+    input::InputStack &inputStack = mwg.inputStack;
+
+    viewStack.pop();
+    inputStack.pop();
+
+    return Var::Empty;
+  }
+};
+
+void registerViewShards() {
+  REGISTER_SHARD("GFX.View", ViewShard);
+  REGISTER_SHARD("GFX.PushView", PushViewShard);
```

Comment:
I don't like Push/Pop with shards.. it's complicated cos you can resume other coroutines and what not in shards. Hard to keep track. One Wire can stop (where you push) while other resumes (where you pop).
If possible I prefer (-> contexts.

Reply:
Yeah fair, I'm not sure why I didn't do that from the start, will change it.